src/estimagic/parameters/nonlinear_constraints.py

```diff
@@ -0,0 +1,468 @@
+            optimization faster, especially for very fast constraint functions. Default
+            False.
+
+    Returns:
+        list[dict]: List of processed constraints.
+
+    """
+    constraint_evals = [
+        _check_validity_and_return_evaluation(c, params, skip_checks)
+        for c in nonlinear_constraints
+    ]
+
+    _processor = partial(
+        _process_nonlinear_constraint,
+        params=params,
+        converter=converter,
+        numdiff_options=numdiff_options,
+    )
+
+    processed = [
+        _processor(c, e) for c, e in zip(nonlinear_constraints, constraint_evals)
+    ]
+    return processed
+
+
+def _process_nonlinear_constraint(
+    c, constraint_eval, params, converter, numdiff_options
+):
+    """Process a single nonlinear constraint."""
+
+    # ==================================================================================
+    # Process selectors for params (functional) and flat_params (indices)
+    # ==================================================================================
+
+    external_selector = _process_selector(c)  # functional selector
+
+    selection_indices, n_params = _get_selection_indices(params, external_selector)
+
+    # ==================================================================================
+    # Evaluate constraint function and selector
+    # ==================================================================================
+    selected = external_selector(params)
+
+    constraint_func = c["func"]
+
+    if constraint_eval is None:
+        constraint_eval = constraint_func(selected)
+
+    # ==================================================================================
+    # Retrieve number of constraints
+    # ==================================================================================
+    if "n_constr" in c:
+        _n_constr = c["n_constr"]
+    else:
+        constraint_value = constraint_func(external_selector(params))
+        _n_constr = len(np.atleast_1d(constraint_value))
+
+    # ==================================================================================
+    # Consolidate and transform jacobian
+    # ==================================================================================
+
+    # process numdiff_options for numerical derivative
+    options = numdiff_options.copy()
+    options.pop("lower_bounds", None)
+    options.pop("upper_bounds", None)
+
+    if "derivative" in c:
+        if not callable(c["derivative"]):
+            msg = "Jacobian of constraints needs to be callable."
+            raise ValueError(msg)
+        jacobian = c["derivative"]
+    else:
+        # use finite-differences if no closed-form jacobian is defined
+        def jacobian(p):
+            return first_derivative(
+                constraint_func,
+                p,
+                **options,
+            )["derivative"]
+
+    def _jacobian_from_internal(x):
```

Comment:
Let's call this `_internal_jacobian`; `_jacobian_from_internal` sounds like it takes some internal jacobian matrix and returns an external one.